docs/content/docs/concepts/Breaking Changes and Supported Patterns.md

```diff
@@ -0,0 +1,293 @@
+
+There are a few approaches that can be taken to resolve this:
+
+1. Restrict the Supported Patterns for the interface. Depending on how its used there are two main choices:
+   1. "out": Disallow implementing and extending the interface and modifying instances. This means the package can add members, remove optionality and make members have more specific types. This means that the old interface must be assignable to the new interface. Uses also should not assume that enumerating members of objects that meet the interface won't return unexpected entries (this is true in general, but extra important for interfaces using these rules). This restricts the instances of the interface to being produced by the package making it logically an "output" of the package.
+   2. "in": Disallow reading members of the interface (except via `{added: value, ...existingInstance}` pattern). This means the package can remove members, add optionality and make members have more general types. This means that the new interface must be assignable to the old interface. This restricts the instances of the interface to being build but not consumed by user code making it logically an "input" of the package.
+2. "Invariant": Create a new interface instead of modifying the existing one. Probably pick one of the above options to apply to the new interface to avoid having to make even more of them.
+
+Interestingly these three cases correspond to [covariant, contravariant and invariant from type systems](<https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)>), including in TypeScript which added [explicit use of them in 4.7](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-7.html#optional-variance-annotations-for-type-parameters).
+
+Just like TypeScript added support for explicitly restricting usage of generic types based on the variance annotations on its type parameters, we can restricted the Supported Patterns of The Package based based on variance annotations we add to its types.
+In our example this could look like:
+
+```typescript
+/**
+ * @out
+ */
+interface Name {
+  first: string;
+  last?: string;
+}
+```
+
+This would make User Code 1 and 2 unsupported. This is supposed to allow changes like:
+
+```typescript
+/**
+ * @out
+ */
+interface Name {
+  first: string;
+  middle: string; // Added. This is supposed to be ok, but can still result in name collisions, like breaking UserCode 3.
+  last: string; // Made required. This is ok.
+}
+```
+
+To solve this issue with User Code 3, we need to decide who owns the unused portion of the namespace on types defined in the package.
+Rather than documenting this for every time, a single default policy should be clearly stated, and exception to it can be documented.
+Defaulting to the package owing the namespace is probably best, since the User Code can just accept that that use is not supported or extend it using private symbols without conflicting if needed.
```

Comment:
```suggestion
Defaulting to the package owning the namespace is probably best, since the User Code can just accept that that use is not supported or extend it using private symbols without conflicting if needed.
```